src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -3592,384 +3607,173 @@ private unsafe void ParseRemaining()
             InterlockedSetFlags(cF);
         }
 
-        //
         // verifies the syntax of the scheme part
         // Checks on implicit File: scheme due to simple Dos/Unc path passed
         // returns the start of the next component  position
-        // throws UriFormatException if invalid scheme
-        //
-        private static unsafe int ParseSchemeCheckImplicitFile(char* uriString, int length,
-            ref ParsingError err, ref Flags flags, ref UriParser? syntax)
+        private static int ParseSchemeCheckImplicitFile(string uriString, ref ParsingError err, ref Flags flags, ref UriParser? syntax)
         {
+            Debug.Assert(err == ParsingError.None);
             Debug.Assert((flags & Flags.Debug_LeftConstructor) == 0);
 
-            int idx = 0;
+            int i = 0;
 
-            //skip whitespace
-            while (idx < length && UriHelper.IsLWS(uriString[idx]))
+            // skip whitespace
+            while ((uint)i < (uint)uriString.Length && UriHelper.IsLWS(uriString[i]))
             {
-                ++idx;
+                i++;
             }
 
             // Unix: Unix path?
             // A path starting with 2 / or \ (including mixed) is treated as UNC and will be matched below
-            if (!OperatingSystem.IsWindows() && idx < length && uriString[idx] == '/' &&
-                (idx + 1 == length || (uriString[idx + 1] != '/' && uriString[idx + 1] != '\\')))
+            if (!OperatingSystem.IsWindows() &&
+                (uint)i < (uint)uriString.Length && uriString[i] == '/' &&
+                ((uint)(i + 1) >= (uint)uriString.Length || uriString[i + 1] is not ('/' or '\\')))
             {
                 flags |= (Flags.UnixPath | Flags.ImplicitFile | Flags.AuthorityFound);
                 syntax = UriParser.UnixFileUri;
-                return idx;
+                return i;
             }
 
-            // sets the recognizer for well known registered schemes
-            // file, ftp, http, https, uuid, etc
+            // Find the colon.
             // Note that we don't support one-letter schemes that will be put into a DOS path bucket
 
-            int end = idx;
-            while (end < length && uriString[end] != ':')
+            int colonIndex = i;
+            while ((uint)colonIndex < (uint)uriString.Length && uriString[colonIndex] != ':')
```

Comment:
Should we use IndexOf, or are you purposefully avoiding it because this is expected to find the ':' immediately?

Reply:
Switched to `IndexOf` now.
For most actual urls the expected match is gonna be close to the start, so perf-wise it's a wash.
But `IndexOf` will do a better job for implicit file paths that also take this path.